include/envoy/registry/registry.h

```diff
@@ -188,6 +189,50 @@ template <class Base> class FactoryRegistry : public Logger::Loggable<Logger::Id
     return *deprecated_factory_names;
   }
 
+  /**
+   * Lazily constructs a mapping from the configuration message type to a factory,
+   * including the deprecated configuration message types.
+   * Must be invoked after factory registration is completed.
+   */
+  static absl::flat_hash_map<std::string, Base*>& factoriesByType() {
+    static absl::flat_hash_map<std::string, Base*>* factoriesByType = [] {
+      auto* mapping = new absl::flat_hash_map<std::string, Base*>();
+
+      for (const auto& factory : factories()) {
+        if (factory.second == nullptr) {
+          continue;
+        }
+
+        // skip untyped factories and factories that consume Struct
```

Comment:
Which factories consume `Struct`?

Reply:
Only test factories AFAICT.

Reply:
It feels kind of bad that we need to have prod code here that exists just for test, any way we can improve that, e.g. switch tests to use some canonical `TestConfigMessage`?

Reply:
I didn't think deeply about filters that use struct as config type. I'd out-right ban Struct but that requires deprecation and will to force everyone to create their own proto types. Can this be done separately? What we did with Empty was easier since it was mostly used in Envoy code base.

Reply:
Done.